nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -102,6 +102,7 @@ Costmap2DROS::Costmap2DROS(const std::string & name, tf2_ros::Buffer & tf)
 
   // we need to make sure that the transform between the robot base frame
   // and the global frame is available
+  rclcpp::Rate r(10);
```

Comment:
I'm not sure this is safe. At this point in the code, the node isn't spinning yet. If it's not spinning, the ros clock can't advance when `use_sim_time` is active. This could deadlock.

It may not be a failure in practice because, last time I checked, `Rate` used the wall clock instead of the ros clock, however that seems like a bug to me and may change. I'd say we should only make this change if we are confident that Rate objects should always use wall clock time and never the simulation clock.

Reply:
From this ticket: https://answers.ros.org/question/319510/rclcpprate-doesnt-use-ros-time-use_sim_time/

I'd venture to say it will not remain that way. However, even if the node hasn't been put in spin, that has never before stopped the ros clock - that's usually been a separate library that can be instantiated separately of even a ros node. I can't think of a reason why the node spinning would stop the clock.

Reply:
It's not that way anymore. The clock updates come in as a message on the /clock topic and the node has to spin to process that message and advance the clock

At least, that's the way it was a few months ago.

Reply:
Got it, OK. I think all 3 need throttling (especially the service caller). I will redo using non-ros clock implementations. 

Thanks for the deep look